### ISSUE TITLE
Change PEP 647 to allow for narrowing in the negative case

### ISSUE THREAD

**Issue by Zomatree** [python/typing#926]
Currently [PEP 647](https://www.python.org/dev/peps/pep-0647/) doesnt narrow the type in the negative case, i feel like this should be changed to narrow in the negative case, currently due to it not narrowing its impossible to correctly type certain functions with typeguards fully, this came up in actual code a couple times, for example:

this example is used in conjuction with https://github.com/python/typeshed/pull/5658
```py
R_T = TypeVar("R_T")
P = ParamSpec("P")

async def maybe_coroutine(func: Callable[P, Union[R_T, Coroutine[Any, Any, R_T]]], *args: P.args, **kwargs: P.kwargs) -> R_T:
    value = func(*args, **kwargs)

    if isawaitable(value):
        value = await value

    return value
```
because `value` isnt narrowed to be `R_T@maybe_coroutine` outside the if statement its back to being a union even though it would never be awaitable because of the if statement, this feels like a massive downside and makes typeguard pretty useless for a lot of cases, forcing people to use `cast` or `# type: ignore`.

While this may cause a lot of controversy with changing a quite large part of how typeguards work, i feel like this is a needed change and a step in the right direction for making typing much more powerful and able to be properly integrated into more projects codebases without having to resort to ignoring or overwrites the type

**Comment by JelleZijlstra** [comment]
cc @erictraut. I know Eric has said before that this was a conscious decision based on experience with TypeScript, but I don't recall the details and PEP 647 doesn't actually cover this in its "Rejected ideas" section. It would be useful to add an explicit rejection of the idea to the PEP.

I've definitely also felt the need for a TypeGuard that narrows in the negative case too.

**Comment by Zomatree** [comment]
I have no experience with typeguards in typescripts so if this is carried over from typescript i would like to know why and maybe potential ways around this or a better solution to this issue

**Comment by gvanrossum** [comment]
IIRC the issue is that a typeguard function may reject objects that match the given type because their value doesn't match. Maybe I can make up an example.
```py
def is_small_int(x: object) -> TypeGuard[int]:
    return isinstance(x) and 0 <= x < 256

def foo(x: int|str) -> str:
    if is_small_int(x):
        return str(x)
    elif isinstance(x, int):
        return hex(x)
    else:
        return repr(x)
```
Here the type checker cannot assume that the elif clause is unreachable just because is_small_int() returns false -- x could still be an int.

**Comment by Zomatree** [comment]
I see this as a non-issue, If you want to check the value of the int in your typeguarded function then the typeguard's value should also reflect that, in this case its perfectly reasonable to define the function to something like:
```py
def is_small_int(x: object) -> TypeGuard[Literal[0, 1, 2, 3, 4, 5, ...]]:
    return isinstance(x) and 0 <= x < 256
```

for more complex checks that could'nt be described in the return type, you should'nt be doing that in a typeguard or if this where a perfectly world where the python's type system was powerful enough then there should'nt be anything that could'nt be described via the return type.

**Comment by erictraut** [comment]
Negative type narrowing cannot be safely applied for user-defined type guards. A user-defined type guard check indicates only that the value is of a given type. It is not type safe to assume that a type can be eliminated from a union if this function returns False.

I'll provide a couple of concrete examples. The first one is a flavor of a problem that bites me periodically in TypeScript, which does provide narrowing in the negative case when the input argument is a union. This is the reason why the TypeScript team suggested that they somewhat regret making the decision.

```python
class Cat:
    color: str

class Dog:
    breed: str

Pet = Cat | Dog

def is_black_cat(val: Pet) -> TypeGuard[Cat]:
    return isinstance(val, Cat) and val.color == "black"

def filter_back_cats(pets: list[Pet]) -> list[Cat]:
    return [x for x in pets if is_black_cat(x)]

def print_pet_info(pet: Pet):
    if is_black_cat(pet):
        print("Found black cat!")
    elif isinstance(pet, Cat):
        # This would be incorrectly flagged as an unreachable piece of code
        # if is_black_cat() eliminated Cat from the union.
        print("Regular cat")
    else:
        print("Not a cat")
```

TypeScript is able to get the "correct" answer most of the time in the negative case because user-defined type guards in TypeScript require that the TypeGuard type (the "narrowed" type) is a proper subtype of the value type (the type of the input parameter). It's able to use this to determine which subtype of the input union type should be eliminated and which should be retained in the negative case.

When we were defining the rules for PEP 647, we decided that this limitation was too constraining for Python, and we relaxed this requirement. This makes the handling of the negative case even more fraught with problems.

Consider this example:

```python
def is_str_list(val: Sequence[int | str | None]) -> TypeGuard[List[str]]:
    return isinstance(val, list) and all(isinstance(x, str) for x in val)

def func1(val: Sequence[str | int] | List[str | None]):
    if is_str_list(val):
        return

    # What is the type of `val` here?
    reveal_type(val)
```

Your assertion that user-defined type guards are "pretty useless for a lot of cases" without negative narrowing strikes me as a bit hyperbolic. In practice, I don't think this limitation is that big of a problem in most cases.

Significant thought and discussion went into this PEP during its design, and we concluded that this was a reasonable tradeoff.

In any event, the change you're proposing would be a breaking change at this point since Python 3.10 is released and people have started to use TypeGuard in its current form and assuming its currently-defined semantics.

**Comment by Zomatree** [comment]
While i can see the issue now, there are definitely ways to fix these issues, the one that comes to my mind is to introduce an Intersection type (https://github.com/python/typing/issues/213) and do something similar to:
```py
class BlackCatProto(Protocol):
    color: Literal["black"]

def is_black_cat(pet: Pet) -> TypeGuard[Cat & BlackCatProto]:
    return isinstance(pet, Cat) and pet.color == "black"
```
this is similar to how typescript does it as well
```ts
function isBlackCat(pet: Pet): pet is Cat & { color: "black" } {
  return pet instanceof Cat && v.color === "black";
}
```

**Comment by hmc-cs-mdrissi** [comment]
Intersection types are complex enough of a feature it is unlikely that supporting narrowing in negative case is reason to add intersection types. I expect intersection types to need there own pep/campaigning to happen along with clear choices as to are intersections nominal/structural/something else?

I think alternative approach for Negative narrowing is having second variant of TypeGuard like ExactTypeGuard[T] where the function being false implies the value is not type T while normal TypeGuards continue to only cover positive case. I mainly care about union negative narrowing logic as that's main time the issue appears for me and currently I work around it with casts as needed.

**Comment by MajorDallas** [comment]
I was also expecting the behavior that, given a binary choice, `TypeGuard` could narrow the type in the positive and negative cases. In my usage, I have a simple binary choice between two possible implementations of a custom abc, with the decision ultimately determined by a `ClassVar[bool]` attribute. Getting Mypy and Pyright to recognize the one option in the positive case but continue to throw errors in the negative case got frustrating pretty quickly.

It's especially frustrating seeing that the motivating examples for this behavior are conflating type-guarding with value-checking. `color == "black"` doesn't change whether `pet` is of **type** `Cat`, and there's no way to statically know if `color` is "black" or "tabby," anyway! But, since `TypeGuard` is implemented to rely on a runtime function call, there's no way to prevent this (absolutely incorrect, imo) usage, and, given that, I have to agree that the safest option _is_ this behavior... 

... but it **almost** defeats the purpose of having `TypeGuard` _at all_, imo. The current behavior is probably useful if you have lots of `if/elif/elif/.../else` stuff going on, as each successive type guard isn't affected by the assumptions of the previous one. Yet, the ostensibly simplest case of a binary choice between statically known options is impossible to type safely (i.e. no abusing `cast`). 

When I got to the point of trying to get the negative path correctly typed, I was hoping I could just pass a second argument to `TypeGuard`:

```
def is_ok_variant(variant: Result[T, E]) -> TypeGuard[Ok[T], Err[E]]:
    return isinstance(variant, Ok)
```

Without strict narrowing, this isn't any better than a `cast`, however.

I like the idea of an Intersection type as described above. I also like the idea of a disjoint union or Sum type. Scala has this as traits, Rust has it as Enums. It's a great complement to pattern matching, which we now have (yay!). I've actually tried coaxing Python's `Enum` into working like a "sum of types" type, even had a tiny bit of success from hacking on `_missing_`, but Mypy doesn't know what to do with it in my experiments so far and it's not ergonomic at all:

```
class ResultEnum(Enum):
    OK = type("Ok", (), {...})
    ERR = type("Err", (), {...})
    def _missing_(....

result = ResultEnum(func_that_returns_OkorErr())
if result is ResultEnum.OK:
    ...
else:
    # If it's known we got some ResultEnum member and that there are only two members of the enum, it can be statically proven that there is exactly one other possibility. In theory.
    ...
```

Unfortunately, although nothing explodes if the value of an `Enum` member is a type, you can't actually instantiate one the usual way, i.e. `ResultEnum.OK(val)` won't work and using `ResultEnum.OK.value(val)` is pretty clunky. Also, in the truncated example above, you'd lose the instance of the type, but there are hacks for that, too. I believe a new but similar metaclass could make it viable, but my experiments in that direction haven't gotten far. 

Anyway, the idea is for the object itself to have a statically-checkable tag of some sort (hence some literature calls Sum/disjoint union types "tagged unions"). Such a tag could be a read-only property that's not tied to an attribute and returns a literal, membership in an Enum, or anything else that couldn't possibly change at runtime. 

With that, it might be possible to have a proper, _static_ type guard solution that can't be abused with runtime attribute checks, maybe even do exhaustive type narrowing of every member of the Sum type.

**Comment by MajorDallas** [comment]
I also found `TypeGuard`'s behavior surprising and unintuitive. For what I needed it for, not being able to narrow in the negative case also meant that I had to find another solution. 

At least for me, the confusion arises from the examples above. In Eric's `Cat` example, for instance:

```python
blackcat = Cat(color="black")
tabbycat = Cat(color="tabby")
reveal_type(blackcat)  # Cat
reveal_type(tabbycat)  # Cat
if not is_black_cat(tabbycat):
    print(type(tabbycat))  # still Cat
```

To me, inspecting the runtime value of `color` is not a "type" guard but a "state" guard. `TypeGuard` actually being a state guard is what made the behavior surprising, and I'm inclined to agree with it being "pretty useless for a lot of cases" in the static analysis context.
